React 18, remove onLoad, add openCvVersion param

erasta opened the pull request:
In this PR i introduced a few needed changes:
1. Advance to React 18: [ReactDOM.render is deprecated](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis), `index.css` breaks.
2. `onLoad` is not needed because `useContext` is called again when when cv is loaded.
3. Param `openCvVersion` is added.
4. `window.Module` is changed only inside `useEffect` and only if cv is not loaded.  

I really appreciate the great work you've done on this package, and just wanted to update it a bit. Hope you accept my changes.

erasta committed:
use react versions higher than 16

erasta committed:
advance to react 18 and use createRoot

erasta committed:
revert package-lock.json

erasta committed:
use context instead of onLoad, added openCvVersion param

erasta committed:
keeping with useOpenCv api, semver higher than

erasta committed:
semver higher on peerDependencies

erasta committed:
revert file to link on package.json

erasta committed:
revert yarn.lock files

giacomocerquone commented on example/package.json:
```diff
@@ -18,9 +18,6 @@
   "devDependencies": {
     "@babel/plugin-syntax-object-rest-spread": "^7.8.3"
   },
-  "eslintConfig": {
-    "extends": "react-app"
-  },
```
why are you suggesting to remove the eslint extension in the example project?

erasta replied:
when running `npm start` without removing the `eslistConfig` we get the follow error:
```
ERROR in [eslint] Plugin "react" was conflicted between "package.json » eslint-config-react-app »
 .../opencv-react/example/node_modules/eslint-config-react-app/base.js" and "BaseConfig »
 .../opencv-react/node_modules/react-scripts/node_modules/eslint-config-react-app/base.js".
```
maybe this relates to not upgrading all the packages, specifically `eslint`, but i did not want to over-extend the reach of this PR...

erasta replied:
fixed and re-added. @giacomocerquone

giacomocerquone commented:
Thank you very much for the work. This pr looks promising! I'll give it a proper review soon

erasta committed:
fixed eslist on example app by linking eslint-config-react-app

giacomocerquone commented on src/lib/OpenCvProvider.js:
```diff
@@ -6,53 +6,39 @@ const { Consumer: OpenCvConsumer, Provider } = OpenCvContext
       return
     }
 
+    window.Module = window.Module || {
+      wasmBinaryFile: 'opencv_js.wasm',
+      usingWasm: true,
+      onRuntimeInitialized: () => {
+        console.log('opencv loaded', window.cv)
+        setCvInstance({ loaded: true, cv: window.cv });
+      }
+    }
+
```
shouldn't we reinstatiate it if opencvpath or opencvversion changes?
And why are you worried about checking window.Module first if this effect only reruns when one of the two changes?

I honestly wouldn't place the instance into the state too

erasta replied:
about reinstantiating: in your original code you exit the `useEffect` if there's already a `window.cv` instance available https://github.com/giacomocerquone/opencv-react/blob/1d15d74346cb2cd583852a9ff9fe895d55cf9a44/src/lib/OpenCvProvider.js#L26 so none of this happens when opencvpath or opencvversion changes, and i'm not sure if loading multiple cv instances is something we want to allow.

erasta replied:
about putting the instance into state, i did some external tests and it doesn't seem to be encumbering the state. if you don't put it into the state, the users of `useOpenCv` actually have to take the instance themselves from `window.cv` which will lead to non-encapsulation of the loading process, and some more checks on their behalf on each use.
and i think it is nicer this way...

giacomocerquone commented on src/lib/OpenCvProvider.js:
```diff
@@ -6,53 +6,39 @@ const { Consumer: OpenCvConsumer, Provider } = OpenCvContext
+  const [cvInstance, setCvInstance] = React.useState({ loaded: false, cv: undefined });
 
   React.useEffect(() => {
+    const scriptId = 'opencv-react'
```
this can be left outside where it was, no need to redefine it at each effect run

erasta replied:
agreed.

giacomocerquone commented on src/lib/OpenCvProvider.js:
```diff
@@ -6,53 +6,39 @@ const { Consumer: OpenCvConsumer, Provider } = OpenCvContext
+      }
+    }
+
     // https://docs.opencv.org/3.4/dc/de6/tutorial_js_nodejs.html
     // https://medium.com/code-divoire/integrating-opencv-js-with-an-angular-application-20ae11c7e217
     // https://stackoverflow.com/questions/56671436/cv-mat-is-not-a-constructor-opencv
```
these comments should be placed above window.module, because they explain why we need to do such a thing

erasta replied:
good, will put them there.

giacomocerquone commented:
Ok, I'm dedicating a little bit more time to open source lately. I'm sorry I didn't get back to you earlier, but this project doesn't get traction at all and I don't use this code anymore, therefore it's hard for me to keep up.

But let's try the both of us to follow the flow and try to bring this home 🙂 
I've also added the hacktoberfest topic so that you can get acknowledged for that: https://hacktoberfest.com/

Let me comment on the stuff you wrote:

1. ok for this, it's just about the example though, so not too many worries. Also ok for the CSS, it's just an empty page with a string (might be useful to enhance this a little bit maybe in the future, but not in this pr).
2. I don't think useContext is called again when cv is loaded. Can you expand on this please?
3. Good idea but I really can't just bump up from 3.x to 4.x man 😅  the code looks ok but let's default the version param to the version that we have right now in master.
4. This is no good, see the comments in the code

giacomocerquone commented on package.json:
```diff
@@ -24,28 +24,29 @@
-    "eslint-plugin-react": "^7.17.0",
-    "eslint-plugin-standard": "^4.0.1",
-    "gh-pages": "^2.2.0",
+    "@babel/eslint-parser": "^7.19.1",
+    "cross-env": "^7.0.3",
+    "eslint": "^8.24.0",
+    "eslint-config-prettier": "^8.5.0",
+    "eslint-config-react-app": "^7.0.1",
+    "eslint-config-standard": "^17.0.0",
+    "eslint-config-standard-react": "^11.0.1",
+    "eslint-plugin-import": "^2.26.0",
+    "eslint-plugin-node": "^11.1.0",
+    "eslint-plugin-prettier": "^4.2.1",
+    "eslint-plugin-promise": "^6.0.1",
+    "eslint-plugin-react": "^7.31.8",
+    "eslint-plugin-standard": "^5.0.0",
+    "gh-pages": "^4.0.0",
+    "microbundle-crl": "^0.13.11",
     "npm-run-all": "^4.1.5",
-    "prettier": "^2.0.4",
-    "react": "^16.13.1",
-    "react-dom": "^16.13.1",
-    "react-scripts": "^3.4.1"
+    "prettier": "^2.7.1",
+    "react": ">=18.2.0",
+    "react-dom": ">=18.2.0",
+    "react-scripts": ">=5.0.1"
```
can you please update the strict necessary for your pr?

In general when contributing, always remember to do the limited amount of work requested to fix/implement a certain issue/feature.

erasta replied:
good points, will try to filter which is needed.

erasta committed:
revert to using moduleConfig

erasta committed:
revert to using generateOpenCvScriptTag function

erasta committed:
updating package to strict except eslint (for now)

erasta committed:
reverted eslint package to strict needed

erasta committed:
reverted default opencv version to 3.4.16

erasta committed:
minor reorg to example package.json

erasta commented:
Thank you very much for your comments, I fixed my changes accordingly.
I Hope it is ok now.

Also answered below to your points:

> Ok, I'm dedicating a little bit more time to open source lately. I'm sorry I didn't get back to you earlier, but this project doesn't get traction at all and I don't use this code anymore, therefore it's hard for me to keep up.
> 
> But let's try the both of us to follow the flow and try to bring this home 🙂 I've also added the hacktoberfest topic so that you can get acknowledged for that: https://hacktoberfest.com/  

Great! do you need to do anything else to activate this?

> 
> Let me comment on the stuff you wrote:
> 
> 1. ok for this, it's just about the example though, so not too many worries. Also ok for the CSS, it's just an empty page with a string (might be useful to enhance this a little bit maybe in the future, but not in this pr).
> 2. I don't think useContext is called again when cv is loaded. Can you expand on this please?

useContext itself is not called again, but as said on react docs:
`A component calling useContext will always re-render when the context value changes`
https://reactjs.org/docs/hooks-reference.html#usecontext
(and since `cv` loading happens only once due to your condition on the `useEffect`, there's no need for memoization)

> 3. Good idea but I really can't just bump up from 3.x to 4.x man 😅  the code looks ok but let's default the version param to the version that we have right now in master.

Totally correct... changed back to 3.x, specifically to 3.4.16, which has some minor bug fixes from 3.4.13 but no breaking changes.

> 4. This is no good, see the comments in the code

fixed that.

giacomocerquone commented on example/package.json:
```diff
@@ -4,10 +4,10 @@
-    "build": "node ../node_modules/react-scripts/bin/react-scripts.js build",
-    "test": "node ../node_modules/react-scripts/bin/react-scripts.js test",
-    "eject": "node ../node_modules/react-scripts/bin/react-scripts.js eject"
+    "start": "react-scripts start",
+    "build": "react-scripts build",
+    "test": "react-scripts test",
+    "eject": "react-scripts eject"
```
with your modification to the package.json of the example, I can't start it anymore: 
![image](https://user-images.githubusercontent.com/9303791/197052157-fd1ebf0a-880c-48f6-993a-2b43b6f9d982.png)

erasta replied:
reverted my change, should be ok now.
this was probably because i started it with `npm start`, not `yarn start`. now both ways work well.

giacomocerquone replied:
weird

erasta committed:
revert run-scripts to use node explicitly

giacomocerquone commented:
nice, thanks for the changes 😄

erasta commented:
Great, thank you.
Are you going to update the `npm` package? it would greatly help.
Any way i could assist in that?

giacomocerquone commented:
> Great, thank you. Are you going to update the `npm` package? it would greatly help. Any way i could assist in that?

yes I'm going to very soon, sorry for the delay!

giacomocerquone commented:
Just published version 0.0.6 https://www.npmjs.com/package/opencv-react
